### PR TITLE
Refactor kindyn

### PR DESCRIPTION
**Pull request by thobotics**
Refactor kindyn with
- Feedback control from magnetic sensors
- Filtering to reject bad measurement

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -101,14 +108,22 @@ target_link_libraries(ForcePositionController ${catkin_LIBRARIES} CardsflowHardw
 
 # kindyn library
 add_library(kindyn SHARED
-		src/robot.cpp
+#        src/robot.cpp
+		src/robot_new.cpp
```
Comment (missxa) [inline CMakeLists.txt]: let's keep the name as robot.cpp

File: include/kindyn/controller/cardsflow_command_interface.hpp [hunk 2]
```diff
@@ -59,7 +59,7 @@ namespace hardware_interface {
          * @param motor_cmd cable command
          */
         CardsflowHandle(const CardsflowStateHandle &js, double *joint_position_cmd, double *joint_velocity_cmd,
-                        double* joint_torque_cmd, VectorXd *motor_cmd);
+                        double* joint_torque_cmd, VectorXd *motor_cmd, double* Kp_dl, double* Kd_dl);
```
Comment (missxa) [inline include/kindyn/controller/cardsflow_command_interface.hpp]: rename Kp_dl to Kp and Kd

add them in the comment above with params

File: include/kindyn/filter/joints_kalmanfilter.h [hunk 3]
```diff
@@ -0,0 +1,82 @@
+//
+// Created by roboy on 07.06.21.
+//
+
+#ifndef ROBOY3_JOINTS_KALMANFILTER_H
+#define ROBOY3_JOINTS_KALMANFILTER_H
+
+#include <ros/ros.h>
+#include <bfl/filter/extendedkalmanfilter.h>
```
Comment (missxa) [inline include/kindyn/filter/joints_kalmanfilter.h]: add to the readme as a dependency and instructions how to build from source

File: src/controller/cableLengthController.cpp [hunk 4]
```diff
@@ -38,6 +38,47 @@
 
 using namespace std;
 
+static const double     _PI= 3.1415926535897932384626433832795028841971693993751058209749445923078164062862089986280348;
```
Comment (missxa) [inline src/controller/cableLengthController.cpp]: why not use constants from existing libs ?

File: src/controller/cableLengthController.cpp [hunk 5]
```diff
@@ -38,6 +38,47 @@
...
+static const double _TWO_PI= 6.2831853071795864769252867665590057683943387987502116419498891846156328125724179972560696;
+
+double mod(double x, double y)
```
Comment (missxa) [inline src/controller/cableLengthController.cpp]: why did you have to implement it on your own?

File: src/controller/cableLengthController.cpp [hunk 6]
```diff
@@ -130,7 +170,7 @@ class CableLengthController : public controller_interface::Controller<hardware_i
         return true;
     }
 private:
-    double Kp = 100, Kd = 0; /// PD gains
+    double Kp = 100, Kd = 0.0; /// PD gains
```
Comment (missxa) [inline src/controller/cableLengthController.cpp]: remove in unused

File: src/kinematics.cpp [hunk 7]
```diff
@@ -0,0 +1,474 @@
...
+//            P.block(6 * k, 6 * a, 3, 3) = link_to_link_transform[k * number_of_links + a] * R_pe.transpose();
+//            Vector3d v = -r_OP + link_to_link_transform[k * number_of_links + a].transpose() * r_OG;
+//            skew << 0, -v(2), v(1),
+//                    v(2), 0, -v(0),
+//                    -v(1), v(0), 0;
+//            P.block(6 * k, 6 * a + 3, 3, 3) = -link_to_link_transform[k * number_of_links + a] * skew;
+//        }
+//    }
+//
+//    counter++;
```
Comment (missxa) [inline src/kinematics.cpp]: remove

File: src/robot_new.cpp [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+        joint_state_sub = nh->subscribe(topic_root+"/sensing/external_joint_states", 1, &Robot::JointState, this);
+    }
+    joint_target_sub = nh->subscribe(topic_root+"control/joint_targets", 100, &Robot::JointTarget, this);
+    controller_type_sub = nh->subscribe("/controller_type", 100, &Robot::controllerType, this);
+//     joint_state_sub = nh->subscribe("/joint_states", 100, &Robot::JointState, this);
+//    floating_base_sub = nh->subscribe(topic_root+"control/floating_base", 100, &Robot::FloatingBase, this);
+//    ik_srv = nh->advertiseService(topic_root+"control/ik", &Robot::InverseKinematicsService, this);
+//    ik_two_frames_srv = nh->advertiseService(topic_root+"control/ik_multiple_frames", &Robot::InverseKinematicsMultipleFramesService, this);
+//    fk_srv = nh->advertiseService(topic_root+"control/fk", &Robot::ForwardKinematicsService, this);
+//    freeze_srv = nh->advertiseService(topic_root+"control/freeze", &Robot::FreezeService, this);
```
Comment (missxa) [inline src/robot_new.cpp]: bring it back

File: src/robot_new.cpp [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+//    interactive_marker_sub = nh->subscribe(topic_root+"control/interactive_markers/feedback",1,&Robot::InteractiveMarkerFeedback, this);
+//    zero_joints_sub = nh->subscribe(topic_root+"control/zero_joints", 1, &Robot::ZeroJoints,this);
```
Comment (missxa) [inline src/robot_new.cpp]: bring this back too

File: src/robot_new.cpp [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+    l_next.setZero();
+    l_target.setZero();
+
+    controller_type.resize(kinematics.number_of_cables, CARDSflow::ControllerType::cable_length_controller);
+
+    /**
+     * Register ROS control
+     */
+    Kp_dl = new double(100.0);
+    Kd_dl = new double(0.0);
```
Comment (missxa) [inline src/robot_new.cpp]: add comments what it is, also rename to make it clear

File: src/robot_new.cpp [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+
+void Robot::update(){
+
+    if (nh->hasParam("k_dt"))
+        nh->getParam("k_dt", k_dt);
+    if (nh->hasParam("Kp_dl"))
+        nh->getParam("Kp_dl", *Kp_dl);
+    if (nh->hasParam("Kd_dl"))
+        nh->getParam("Kd_dl", *Kd_dl);
+
```
Comment (missxa) [inline src/robot_new.cpp]: maybe add a flag that would enable reading this params, something like `debug` maybe? that it defined in the launch file

Comment (missxa) [reply src/robot_new.cpp]: too avoid reading these parameters all the time

File: src/robot_new.cpp (hunk 10 above)
Comment (missxa) [reply src/robot_new.cpp]: declare default values in the launch file

File: src/robots/upper_body.cpp [hunk 12]
```diff
@@ -239,12 +260,24 @@ class UpperBody: public cardsflow::kindyn::Robot{
...
         vector<float> _integral(motor_ids.size(), 0);
+        vector<float> _error(motor_ids.size(), 0);
         integral[name] = _integral;
+        error_last[name] = _error;
 
+        update();
 
-//        }
+        // Set current state to bullet
+        publishTarget(name, false);
```
Comment (missxa) [inline src/robots/upper_body.cpp]: rename to someting like publishBulletTarget or publishSimulationTarget

File: src/robots/upper_body.cpp [hunk 13]
```diff
@@ -239,12 +260,24 @@ class UpperBody: public cardsflow::kindyn::Robot{
...
+            ROS_INFO_THROTTLE(1, "waiting %d for setting bullet", seconds--);
+        }
+
+        // Move back to zero position
+        publishTarget(name, true);
 
-        update();
         ROS_INFO_STREAM("%s pose init done" << name);
         init_called[name] = true;
         nh->setParam("initialized", init_called);
```
Comment (missxa) [inline src/robots/upper_body.cpp:289]: "initialized" rosparam is a dictionary, so this prob doesn't work

File: src/robots/upper_body.cpp [hunk 14]
```diff
@@ -254,6 +287,44 @@ class UpperBody: public cardsflow::kindyn::Robot{
 
     }
 
+    void publishTarget(string body_part, bool zeroes){
```
Comment (missxa) [inline src/robots/upper_body.cpp]: add an inline comment what this does, it's unclear by the function signature what is actually happening

**Review by missxa — CHANGES_REQUESTED** [review]

File: src/kinematics.cpp (hunk 7 above)
Comment (thobotics) [reply src/kinematics.cpp]: Not resolved yet!

File: src/controller/cableLengthController.cpp (hunk 6 above)
Comment (thobotics) [reply src/controller/cableLengthController.cpp]: Not resolved yet!

File: src/robot_new.cpp (hunk 10 above)
Comment (thobotics) [reply src/robot_new.cpp]: Not resolved yet!

File: src/robot_new.cpp (hunk 9 above)
Comment (thobotics) [reply src/robot_new.cpp]: Not resolved yet!

File: src/robot_new.cpp (hunk 8 above)
Comment (thobotics) [reply src/robot_new.cpp]: Not resolved yet!

File: CMakeLists.txt [hunk 15]
```diff
@@ -101,14 +102,16 @@ target_link_libraries(ForcePositionController ${catkin_LIBRARIES} CardsflowHardw
...
+		src/kinematics.cpp
 		src/EigenExtension.cpp
 		include/kindyn/EigenExtension.hpp
 		include/kindyn/robot.hpp
+		include/kindyn/kinematics.hpp
 		include/kindyn/cable.hpp)
 add_dependencies(kindyn roboy_simulation_msgs_generate_messages_cpp roboy_simulation_msgs_msgs_generate_messages_cpp)
 target_link_libraries(kindyn ${catkin_LIBRARIES} ${iDynTree_LIBRARIES} ${qpOASES_LIBRARIES} CardsflowHardwareInterface
-		CableLengthController TorquePositionController ForcePositionController ${OpenMP_CXX_libraries})
+		CableLengthController TorquePositionController ForcePositionController ${OpenMP_CXX_libraries} ${BFL_LIBRARIES})
```
Comment (missxa) [inline CMakeLists.txt:114]:
```suggestion
		CableLengthController TorquePositionController ForcePositionController ${OpenMP_CXX_libraries} )
```

File: src/robot.cpp [hunk 16]
```diff
@@ -11,269 +15,120 @@ Robot::Robot() {
...
-                        << mdlLoader.model().toString());
-        return;
-    }
-
-    ROS_INFO_STREAM(kinDynComp.getDescriptionOfDegreesOfFreedom());
+    updatePublishers();
+    fmt = Eigen::IOFormat(4, 0, " ", ";\n", "", "", "[", "]");
+    nh->getParam("external_robot_target", external_robot_target);
+    nh->getParam("external_robot_state", external_robot_state);
+    nh->setParam("vr_puppet",false);
```
Comment (missxa) [inline src/robot.cpp:54]: this is one is unused

File: src/robot.cpp [hunk 17]
```diff
@@ -11,269 +15,120 @@ Robot::Robot() {
...
+    // PD gains for cable length controller
+    Kp_ = new double(100.0);
+    Kd_ = new double(0.0);
+    if (nh->hasParam("k_dt"))
+        nh->getParam("k_dt", k_dt);
+    if (nh->hasParam("Kp"))
+        nh->getParam("Kp", *Kp_);
+    if (nh->hasParam("Kd"))
+        nh->getParam("Kd", *Kd_);
+
```
Comment (missxa) [inline src/robot.cpp:102]: we might need different gains for different joints, what's the approach then?

File: src/robots/upper_body.cpp [hunk 18]
```diff
@@ -239,12 +263,27 @@ class UpperBody: public cardsflow::kindyn::Robot{
...
+            publishBulletTarget(name, "current");
+
+            t0 = ros::Time::now();
+            int seconds = 3;
+            while ((ros::Time::now() - t0).toSec() < 3) {
+                ROS_INFO_THROTTLE(1, "waiting %d for setting bullet", seconds--);
+            }
+
+            // Move back to zero position
+            publishBulletTarget(name, "zeroes");
```
Comment (missxa) [inline src/robots/upper_body.cpp:284]: having a string here is not great. at least an enum or a const string which is not pretty either

File: src/robots/upper_body.cpp [hunk 19]
```diff
@@ -254,6 +293,49 @@ class UpperBody: public cardsflow::kindyn::Robot{
 
     }
 
+    /**
+     * Publish Target point to bullet
+     * @param body_part
+     * @param zeroes_or_current will publish either "zeroes" or "current" as targets to Bullet
+     */
+    void publishBulletTarget(string body_part, string zeroes_or_current){
```
Comment (missxa) [inline src/robots/upper_body.cpp:301]: see previous comment - avoid strings due to easy typos occurrence

File: CMakeLists.txt (hunk 15 above)
Comment (thobotics) [reply CMakeLists.txt:114]: Resolved in `fix_L_matrix` branch

File: src/robot.cpp (hunk 16 above)
Comment (thobotics) [reply src/robot.cpp:54]: Resolved in `fix_L_matrix` branch

File: src/robots/upper_body.cpp (hunk 18 above)
Comment (thobotics) [reply src/robots/upper_body.cpp:284]: Resolved in `fix_L_matrix` branch

File: src/robots/upper_body.cpp (hunk 19 above)
Comment (thobotics) [reply src/robots/upper_body.cpp:301]: Resolved in `fix_L_matrix` branch

**Pull request merged**